force-app/nksAddresses/classes/NKS_AddressController.cls

```diff
@@ -1,364 +1,284 @@
+                    latestAddress = addr;
+                }
+            }
+        }
+        return latestAddress;
+    }
+
+    @TestVisible
+    private static PDL_Oppholdsadresse setLastUpdatedOppholdsAdresse(List<PDL_Oppholdsadresse> addresses) {
+        PDL_Oppholdsadresse latestAddress;
         for (PDL_Oppholdsadresse addr : addresses) {
-            if (latestAddress == null) {
+            if (addr.gyldigFraOgMed != null && (latestAddress == null || addr.gyldigFraOgMed > latestAddress.gyldigFraOgMed)) {
                 latestAddress = addr;
-            } else {
-                if (getCreatedDate(latestAddress.metadata) >= getCreatedDate(addr.metadata))
-                    continue;
+            }
+        }
+        if (latestAddress == null) {
+            for (PDL_Oppholdsadresse addr : addresses) {
+                if (latestAddress == null || getCreatedDate(latestAddress.metadata) < getCreatedDate(addr.metadata)) {
+                    latestAddress = addr;
+                }
+            }
+        }
+        return latestAddress;
+    }
+
+    @TestVisible
+    private static PDL_DeltBosted setLastUpdatedDeltBostedsAddress(List<PDL_DeltBosted> addresses) {
+        PDL_DeltBosted latestAddress;
+        for (PDL_DeltBosted addr : addresses) {
+            if (addr.startdatoForKontrakt != null && (latestAddress == null || latestAddress.startdatoForKontrakt < addr.startdatoForKontrakt)) {
                 latestAddress = addr;
             }
         }
+        if (latestAddress == null) {
+            for (PDL_DeltBosted addr : addresses) {
+                if (latestAddress == null || getCreatedDate(latestAddress.metadata) < getCreatedDate(addr.metadata)) {
+                    latestAddress = addr;
+                }
+            }
+        }
         return latestAddress;
     }
 
+    @TestVisible
+    private static void setAddressFullNameAndCity(List<NKS_Address> addresses, Person__c person) {
+        if (addresses == null || person == null) {
+            return;
+        }
+        Set<String> zipCodes = new Set<String>();
+        for (NKS_Address address : addresses) {
+            address.fullName = person.NKS_Full_Name__c?.toUpperCase();
+            if (address.zipCode != null && address.region == null) {
+                zipCodes.add(address.zipCode);
+            }
+        }
+        if (!zipCodes.isEmpty()) {
+            Map<String, Common_Code__c> zipCodeToCityMap = new Map<String, Common_Code__c>();
+            for (Common_Code__c cc : [SELECT CRM_Code__c, Name FROM Common_Code__c WHERE CRM_Code_Set__c = 'Postnummer' AND CRM_Code__c IN :zipCodes]) {
+                zipCodeToCityMap.put(cc.CRM_Code__c, cc);
+            }
+            for (NKS_Address address : addresses) {
+                if (zipCodeToCityMap.containsKey(address.zipCode)) {
+                    address.city = zipCodeToCityMap.get(address.zipCode).Name;
+                }
+            }
+        }
+    }
```

Comment:
Ikke særlig glad i denne, den looper litt for mye unødvendig. Enten så syntes jeg at vi skal query på alle postnummer og gjøre alt i første loopen aka. slette den andre delen her. Og da kunne vi flytte common code mappen opp til en static final variabel med en setter.
Eller så redusere scopet i den innerste for loopen med å f.eks. legge addressene som skal endres i en ny liste i første loopen.

Reply:
Tror ikke jeg helt forstår, du må nesten vise. Det er ingen nested loops her, alt er O(n) x3. Det vil koste mer performance-wise å hente alle records enn å kjøre én ekstra O(n) for-loop